### ISSUE TITLE
Add more slang to Java: for-yield expression

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#105]
Ideally there would be a `for { ... } yield { ... }` expression. 'for' is a keyword, so `Expression.for(...).yield(...)` (or short: `for(...).yield(...)`) does not work. Alternatively `For.of(...).yield(...)` could be used.

Similar to `Property.forAll(...).suchThat(...).check()`, the expression arguments are internally boxed to function parameters for further processing:

```java
final R result = For.of(expr_1, ..., expr_n).yield((T1 name_1, ..., Tn name_n) -> retVal);

final R result = For.of(
    expr_1,
    ...,
    expr_n
).yield((T1 name_1, ..., Tn name_n) -> {
	...
})

interface For {

    static <T1, R> For1<T1, R> of(T1 expr1) { ... }
    static <T1, T2, R> For2<T1, T2, R> of(T1 expr1, T2 expr2) { ... }
    // etc.

    static class For1<T1, R> {
        public R yield(Function1<T1, R> f1) {
            // ...
        }
    }
    // etc.
}
```

Expressions sent to 'for' may be treated differently by type:

* `java.lang.Iterable<T>` - iterate to get values
* `javaslang.algebra.Monad<T>` - use map / flatMap to yield a value (is it possibly to express this internally?)
* etc.
* by default use value _as is_

Caveat: The type parameter `<T>` is not accessible due to type erasure...

**Comment by danieldietrich** [comment]
like so http://docs.scala-lang.org/tutorials/FAQ/yield.html

**Comment by lukaseder** [comment]
That was quick ;-)

**Comment by danieldietrich** [comment]
Yes :) Thanks for sharing

**Comment by danieldietrich** [comment]
Duplicate to #68 , fixed with #1128
